File: sdk/python/fluxdb.py
"""Dependency-free FluxDB HTTP client. Timestamps are decimal nanosecond strings."""
import json
from urllib.request import Request, urlopen
from urllib.parse import quote, urlencode
from urllib.error import HTTPError

class FluxDBError(RuntimeError):
    def __init__(self, status, message):
        super().__init__(message)
        self.status = status

class FluxDBClient:
    def __init__(self, url="http://127.0.0.1:8086", token=None, timeout=30):
        self.url, self.token, self.timeout = url.rstrip("/"), token, timeout

    def request(self, method, path, body=None):
        headers = {"Content-Type": "application/json"}
        if self.token:
            headers["Authorization"] = "Bearer " + self.token
        data = None if body is None else json.dumps(body, allow_nan=False).encode("utf-8")
        try:
            with urlopen(Request(self.url + path, data=data, headers=headers, method=method), timeout=self.timeout) as response:
                raw = response.read()
                return json.loads(raw) if raw else None
        except HTTPError as error:
            raw = error.read().decode("utf-8", errors="replace")
            try:
                message = json.loads(raw).get("error", raw)
            except (ValueError, AttributeError):
                message = raw
            raise FluxDBError(error.code, message) from error

    @staticmethod
    def path(database):
        return "/api/v1/databases/" + quote(database, safe="")

    def databases(self): return self.request("GET", "/api/v1/databases")
    def create_database(self, database): return self.request("POST", self.path(database))
    def drop_database(self, database): return self.request("DELETE", self.path(database))
    def write(self, database, points): return self.request("POST", self.path(database) + "/points", {"points": points})
    def read(self, database, **filters): return self.request("GET", self.path(database) + "/points?" + urlencode(filters))
    def query(self, database, query): return self.request("POST", self.path(database) + "/query", {"query": query})
    def delete(self, database, measurement, start, end, tags=None, exact=False):
        return self.request("DELETE", self.path(database) + "/points", {"measurement": measurement, "start": str(start), "end": str(end), "tags": tags or {}, "exact": exact})
    def schema(self, database): return self.request("GET", self.path(database) + "/schema")
    def retention(self, database): return self.request("GET", self.path(database) + "/retention")
    def set_retention(self, database, seconds): return self.request("PUT", self.path(database) + "/retention", {"seconds": seconds})
    def flush(self, database): return self.request("POST", self.path(database) + "/flush")
    def compact(self, database): return self.request("POST", self.path(database) + "/compact")
    def export(self, database): return self.request("GET", self.path(database) + "/export")

File: sdk/javascript/fluxdb.mjs
/** FluxDB HTTP client for Node.js 20+ and modern browsers. */
export class FluxDBError extends Error {
  constructor(status, message) { super(message); this.name = 'FluxDBError'; this.status = status; }
}
export class FluxDBClient {
  constructor(url = 'http://127.0.0.1:8086', token, timeout = 30000) {
    this.url = url.replace(/\/$/, ''); this.token = token; this.timeout = timeout;
  }
  async request(method, path, body) {
    const response = await fetch(this.url + path, {
      method, headers: { 'Content-Type': 'application/json', ...(this.token ? { Authorization: `Bearer ${this.token}` } : {}) },
      body: body === undefined ? undefined : JSON.stringify(body), signal: AbortSignal.timeout(this.timeout),
    });
    const raw = await response.text();
    let value; try { value = raw ? JSON.parse(raw) : null; } catch { throw new FluxDBError(response.status, raw || 'Invalid server response'); }
    if (!response.ok) throw new FluxDBError(response.status, value?.error || `HTTP ${response.status}`);
    return value;
  }
  path(database) { return '/api/v1/databases/' + encodeURIComponent(database); }
  databases() { return this.request('GET', '/api/v1/databases'); }
  createDatabase(db) { return this.request('POST', this.path(db)); }
  dropDatabase(db) { return this.request('DELETE', this.path(db)); }
  write(db, points) { return this.request('POST', this.path(db) + '/points', { points }); }
  read(db, filters = {}) { return this.request('GET', this.path(db) + '/points?' + new URLSearchParams(filters)); }
  query(db, query) { return this.request('POST', this.path(db) + '/query', { query }); }
  delete(db, measurement, start, end, tags = {}, exact = false) { return this.request('DELETE', this.path(db) + '/points', { measurement, start: String(start), end: String(end), tags, exact }); }
  schema(db) { return this.request('GET', this.path(db) + '/schema'); }
  retention(db) { return this.request('GET', this.path(db) + '/retention'); }
  setRetention(db, seconds) { return this.request('PUT', this.path(db) + '/retention', { seconds }); }
  flush(db) { return this.request('POST', this.path(db) + '/flush'); }
  compact(db) { return this.request('POST', this.path(db) + '/compact'); }
  export(db) { return this.request('GET', this.path(db) + '/export'); }
}
